File: mini_46.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <vector>
#include <stdexcept>

using namespace std;


struct Node {
    string terminal;
    int invalidation_counter;
    unordered_map<char, Node *> prefix_go;
    unordered_map<char, Node *> suffix_go;
    Node *parent;
    Node *sufflink;
    char from_parent;
    vector<string> all_patterns; // only for root

    Node(int total_patterns, Node *p, char fp) {
        terminal = "";
        invalidation_counter = total_patterns;
        from_parent = fp;
        parent = p;
        sufflink = nullptr;
        prefix_go = unordered_map<char, Node *>();
        suffix_go = unordered_map<char, Node *>();
        all_patterns = vector<string>();
    }
};

Node *suf(Node *v, Node *root);
Node *next(Node *v, char c, Node *root);

void invalidate_node(Node *v, int total_patterns) {
    if (v->invalidation_counter < total_patterns) {
        v->invalidation_counter = total_patterns;
        v->suffix_go.clear();
        v->sufflink = nullptr;
    }
}

Node *suf(Node *v, Node *root) {
    invalidate_node(v, root->invalidation_counter);

    if (v->sufflink == nullptr) {
        if (v == root || v->parent == root)
            v->sufflink = root;
        else
            v->sufflink = next(suf(v->parent, root), v->from_parent, root);
    }
    return v->sufflink;
}

Node *next(Node *v, char c, Node *root) {
    invalidate_node(v, root->invalidation_counter);

    if (v->prefix_go.contains(c))
        return v->prefix_go[c];

    if (v == root)
        return root;

    if (v->suffix_go.contains(c))
        return v->suffix_go[c];

    v->suffix_go[c] = next(suf(v, root), c, root);
    return v->suffix_go[c];
}


void add_new_pattern(Node *root, string pattern) {
    invalidate_node(root, root->invalidation_counter + 1);

    Node *current_node = root;
    for (auto c : pattern) {
        if (!current_node->prefix_go.contains(c))
            current_node->prefix_go[c] = new Node(root->invalidation_counter, current_node, c);
        current_node = current_node->prefix_go[c];
    }

    if (current_node->terminal.size() > 0) {
        throw runtime_error("Found 2 exact similar patterns");
    }

    current_node->terminal = pattern;
    root->all_patterns.push_back(pattern);
}


void count_sufflink(Node *current_node, Node *root, unordered_map<string, int> &counter) {

    while (current_node != root) {
        if (current_node->terminal.size() > 0)
            counter[current_node->terminal]++;

        current_node = suf(current_node, root);
    }
}


void findall(Node *root, string str) {

    unordered_map<string, int> counter;
    for (auto &s : root->all_patterns) {
        counter[s] = 0;
    }

    Node *current_node = root;
    for (auto c : str) {
        current_node = next(current_node, c, root);
        count_sufflink(current_node, root, counter);
    }

    for (auto &pair : counter)
        cout << pair.first << ": " << pair.second << '\n';
}


signed main() {

    Node *root = new Node(0, nullptr, 0);

    add_new_pattern(root, "aab");
    add_new_pattern(root, "abba");
    add_new_pattern(root, "bbbab");


    string text = "bbbaabbbbabbbabba";
    findall(root, text);

    cout << "====================\n";
    add_new_pattern(root, "bbab");
    add_new_pattern(root, "bba");
    findall(root, text);

    return 0;
}

File: mini_46.py

class Node:
	def __init__(self, total_patterns: int, parent, from_parent: str):
		self.terminal = ""
		self.invalidation_counter = total_patterns
		self.prefix_go = {}
		self.suffix_go = {}

		self.parent = parent
		self.from_parent = from_parent
		self.sufflink = None
		self.all_patterns = []


def suf(v, root):

	if (v.sufflink is None):
		if v is root or v.parent is root:
			v.sufflink = root
		else:
			v.sufflink = next(suf(v.parent, root), v.from_parent, root)

	return v.sufflink

def next(v, c, root):
	if (c in v.prefix_go):
		return v.prefix_go[c]

	if v is root:
		return root

	if c in v.suffix_go:
		return v.suffix_go[c]

	v.suffix_go[c] = next(suf(v, root), c, root)
	return v.suffix_go[c]


def add_new_pattern(root, pattern: str):
	current_node = root

	for c in pattern:
		if (c not in current_node.prefix_go):
			current_node.prefix_go[c] = Node(root.invalidation_counter, current_node, c)
		current_node = current_node.prefix_go[c]

	assert len(current_node.terminal) == 0

	current_node.terminal = pattern
	root.all_patterns.append(pattern)


def count_sufflink(node, counter, root):

	while node != root:
		if node.terminal:
			counter[node.terminal] += 1

		node = suf(node, root)

def findall(root, string):

	counter = {}
	for s in root.all_patterns:
		counter[s] = 0

	current_node = root
	for c in string:
		current_node = next(current_node, c, root)
		count_sufflink(current_node, counter, root)
		# if current_node.terminal:
		# 	counter[current_node.terminal] += 1


	for key, value in counter.items():
		print(f'{key}: {value}')



def main():
	root = Node(0, None, 0)

	add_new_pattern(root, "aab")
	add_new_pattern(root, "abba")
	add_new_pattern(root, "bbab")
	add_new_pattern(root, "bbbab")

	# check_suff_link(root, "bbbab")

	text = "bbbaabbbbab"
	findall(root, text)
	print('bbab' in text)

if __name__ == '__main__':
	main()

